content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +853,124 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
+```
+node {
+  withCredentials([usernamePassword(credentialsId: 'creds', usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')] {
+    /* CORRECT */
+    sh 'curl -u $USERNAME:$PASSWORD https://example.com/'
+  }
+}
+```
+
+
+====
+
+[WARNING]
+.Injection via interpolation
+====
+
+Another note of caution. Using Groovy string interpolation for user-controlled variables with steps that pass their arguments to command interpreters such as the `sh`, `bat`, `powershell`, or `pwsh` steps can result in problems analogous to SQL injection.
+This occurs when a user-controller variable (generally an environment variable, usually a parameter passed to the build) that contains special characters (e.g. `' " / \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps using Groovy interpolation.
+For a simple example:
+```
+node {
+  withCredentials([string(credentialsId: 'sec-ret-with-symbols', variable: 'SECRET_VALUE')]) {
+    // value of credential = sec>ret
+    sh "echo $SECRET_VALUE"
+  }
+}
+```
+In this example, the argument to the `sh` step is evaluated by Groovy and `SECRET_VALUE` is interpolated directly into the argument as if `sh('echo sec>ret')` has been written in the Pipeline.
+When this is processed on the agent, rather than echoing the value `sec>ret`, it will create a file called `ret` containing the value `sec`.
+Any user able to control a variable interpolated by such a step would be able to make the `sh` step run arbitrary code on the agent.
+To avoid this problem, make sure arguments to steps such as `sh` or `bat` that reference parameters or other user-controller environment variables use single quotes to avoid Groovy interpolation.
+```
+node {
+  withCredentials([string(credentialsId: 'sec-ret-with-symbols', variable: 'SECRET_VALUE')]) {
+    // value of credential = sec>ret
+    sh 'echo $SECRET_VALUE'
+  }
+}
+```
+
+This kind of issue also sometimes occurs accidentally when credentials that contain special characters are passed to a step using Groovy interpolation.
+This can result in the credentials being mangled, causing them to be invalid, and preventing them from being automatically masked in the console log.
+```
+node {
+  withCredentials([string(credentialsId: 'sec-ret-with-symbols', variable: 'SECRET_VALUE')]) {
+    // value of credential = sec%ret
+    bat "echo $SECRET_VALUE"
+  }
+}
+```
+Here, the `bat` step receives `echo sec%ret` and the Windows batch shell will simply drop the `%` and print out the value `secret`.
+Because there is a single character difference, the value `secret` will not be masked.
+Though the value is not exactly the same as the actual password, this is still a significant exposure of sensitive information.
+Again, single-quotes avoids this issue.
+```
+node {
+  withCredentials([string(credentialsId: 'sec-ret-with-symbols', variable: 'SECRET_VALUE')]) {
+    // value of credential = sec%ret
+    bat 'echo $SECRET_VALUE'
+  }
+}
+```
+
+====
+
+
```

Comment:
```suggestion
```

Reply:
That line actually marks the end of the `Injection via interpolation` admonition, so if I remove it, it will go past into the next section.